Add "report formatting for text under braille cell" command

dkager opened the pull request:
### Link to issue number:
Fixes #7106. Fixes #7869.

### Summary of the issue:
This command is useful to obtain formatting information for text under a particular braille cell without first having to route the cursor there. It is an intuitive command to be tied to the Alva BC6 secondary routing keys. Displays without such keys can use a gesture that consists of the routing key with another braille display key.

### Description of how this pull request fixes the issue:
Followed suggestions from @jcsteh in https://github.com/nvaccess/nvda/issues/7106#issuecomment-302020463

Also incorporated PR #7990.

### Testing performed:
Assigned the command to the Alva BC6 secondary routing keys. Then ran it on both Windows controls and text in various editors (Notepad, Word, Notepad++).

### Known issues with pull request:
This command only makes sense when assigned to (secondary) cursor routing keys, because it reports formatting for text under a specific braille cell. However, this is not enforced in any way.

### Change log entry:
In New Features:

- Added a script to report formatting information for the text under a specific braille cell. (#7106)

Changes:

- The secondary routing buttons on Alva BC6, EuroBraille and Papenmeier braille displays now report formatting information for the text under the braille cell of that button. (#7106)

Bug Fixes:

- Pressing the report formatting script no longer moves the reviewPosition and therefore pressing it multiple times no longer gives different results. (#7869)

dkager committed:
Add a "report formatting info for text under this braille cell" command.

dkager committed:
Assign the Alva BC6 secondary cursor routing keys to the new report formatting command.

derekriemer commented:
For other displays, what about tapping adjacent buttons?

On Fri, May 19, 2017 at 4:00 AM, Davy Kager <notifications@github.com>
wrote:

> Link to issue number:
>
> Add "report formatting under cursor routing key" command #7106
> <https://github.com/nvaccess/nvda/issues/7106>
> Summary of the issue:
>
> This command is useful to obtain formatting information for text under a
> particular braille cell without first having to route the cursor there. It
> is an intuitive command to be tied to the Alva BC6 secondary routing keys.
> Displays without such keys can use a gesture that consists of the routing
> key with another braille display key.
> Description of how this pull request fixes the issue:
>
> Followed suggestions from @jcsteh <https://github.com/jcsteh> in #7106
> (comment)
> <https://github.com/nvaccess/nvda/issues/7106#issuecomment-302020463>
> Testing performed:
>
> Tied the command to the Alva BC6 secondary keys. Then ran it on both
> Windows controls and text in various editors (Notepad, Word, Notepad++).
> Known issues with pull request:
>
> The command has only been added to the Alva BC6 keymap. @leonardder
> <https://github.com/leonardder> has offered to add and test other
> displays in a separate PR.
> Change log entry:
>
> In New Features:
>
> - Added a command to report formatting information for the text under a specific braille cell. (#7106)
>
> In Changes:
>
> - The secondary routing buttons on Alva BC6 braille displays now report formatting information for the text under that button. (#7106)
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/nvaccess/nvda/pull/7189
> Commit Summary
>
>    - Add a "report formatting info for text under this braille cell"
>    command.
>    - Assign the Alva BC6 secondary cursor routing keys to the new report
>    formatting command.
>
> File Changes
>
>    - *M* source/braille.py
>    <https://github.com/nvaccess/nvda/pull/7189/files#diff-0> (33)
>    - *M* source/brailleDisplayDrivers/alvaBC6.py
>    <https://github.com/nvaccess/nvda/pull/7189/files#diff-1> (1)
>    - *M* source/globalCommands.py
>    <https://github.com/nvaccess/nvda/pull/7189/files#diff-2> (26)
>
> Patch Links:
>
>    - https://github.com/nvaccess/nvda/pull/7189.patch
>    - https://github.com/nvaccess/nvda/pull/7189.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7189>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivUaW-Sh6efKUWP0QVocrU-Wuiwwrks5r7Wg5gaJpZM4NgUrH>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

   - University of Colorado Boulder Department of computer science, 4th
   year undergraduate student.
   - Accessibility enthusiast.
   - Proud user of the NVDA screen reader.
   - Open source enthusiast.
   - Skier.

Personal website <http://derekriemer.com>

dkager commented:
@leonardder suggested space+routing.

leonardder commented:
> @leonardder suggested space+routing.

Yup, but this won't work with displays without a braille keyboard, Baum Vario 340 for example, or whatever their newest device is called. Also, the 80 cell Braillant displays do not have braille input from what I recall.

dkager commented:
> Also, the 80 cell Braillant displays do not have braille input from what I recall.

... And on the BI 40 I turn it off to avoid bogus input. But I think for anything else, like C1+routing on the Brailliant, users have to decide and define it themselves.

dkager committed:
Update copyrights.

dkager committed:
Merge branch 'master' into i7106

dkager committed:
Update user guide for the new Alva BC6 key assignment.

leonardder commented:
I just walked through this code and all looks good to me.

dkager committed:
Merge branch 'master' into i7106

dkager commented:
Just expertly dealt with a conflict situation, if I say so myself. :)

LeonarddeR reviewed (changes requested):
Ugh, merge conflicts again.

Also, I think it would be good if you would change the papenmeier drivers in this pr as well. See script_upperRouting in both of them.

derekriemer commented:
still merge conflicts

dkager committed:
Merge branch 'master' into i7106

dkager committed:
Also use the new report formatting braille command in both Papenmeier drivers.
This removes their "route to and report formatting" behavior. Instead, it now only reports formatting without routing the cursor.

dkager commented:
All fixed.

michaelDCurran committed:
Merge branch 'master' into i7106

michaelDCurran committed:
Merge branch 'master' into i7106

leonardder commented:
@michaelDCurran: Could it be that you forgot to incubate these?

Nevertheless, there are merge conflicts due to the new ALVA driver.

michaelDCurran commented:
Looks like I just forgot. Weird.

Assuming the merge conflicts are fixed, I'd still wait a bit for Next to 
settle down with the updater changes, then I will incubate this.

dkager committed:
Merge branch 'master' into i7106

dkager committed:
Merge branch 'i7106' of github.com:dkager/nvda into i7106

leonardder commented on source/globalCommands.py:
```diff
@@ -1279,7 +1279,7 @@ def script_sayAll(self,gesture):
 	script_sayAll.category=SCRCAT_SYSTEMCARET
 
-	def script_reportFormatting(self,gesture):
+	def _reportFormatting(self, info, browseable=False):
```
May be call this _reportFormattingHelper or _formatReportingHelper?

leonardder commented on user_docs/en/userGuide.t2t:
```diff
@@ -1927,6 +1927,7 @@ Please see the display's documentation for descriptions of where these keys can
 | Move braille display to next line | t4 |
 | Scroll braille display forward | t5 or etouch3 |
 | Route to braille cell | routing |
+| Reports formatting under braille cell | secondRouting |
```
Given the verbs above are toggle and route, I'd say this should be report instead of report.

leonardder commented on source/globalCommands.py:
```diff
@@ -1907,6 +1913,16 @@ def script_braille_routeTo(self, gesture):
+			# Translators: Reported when trying to obtain formatting information (such as font name, indentation and so on) but there is no formatting information for the text under cursor.
+			ui.message(_("No formatting information"))
+			return
+		self._reportFormatting(info, False)
```
I wonder whether reporting formatting info in a browsable message would make sense for routing with a braille display, e.g. after two or three presses?

dkager committed:
Address review actions.

dkager commented:
@leonardder One more round of review please.

dkager committed:
Merge branch 'master' into i7106

dkager committed:
Update user docs.

dkager committed:
Add the report text formatting under braille cell command to the EuroBraille displays.

LeonarddeR reviewed (changes requested):
Woops, my bad, I overlooked an important thing. When you report formatting under a braille cell, the formatting information is also shown in braille, for the first press, and than the second press applies to the formatting message and not to the text you're trying to review. I guess it's better to revert the double press functionality.

dkager commented:
Interesting.
I also realized that it doesn't make sense to assign this command to a non-routing key, but that we can't enforce that right now.

dkager committed:
Revert browseable message for the report text formatting under braille cell command.

dkager committed:
Merge branch 'master' into i7106

leonardder commented:
This is likely to conflict with #7990

michaelDCurran commented:
Above the info.expand in _reportFormattingHelper, please at the following line:
```
info=info.copy()
```
This ensures that we don't accidentally mutate the passed in TextInfo, which may have been the review cursor or something.
This will then superseed PR #7990 and therefore fix #7869

dkager committed:
Incorporate PR #7990.

dkager commented:
Updated https://github.com/nvaccess/nvda/pull/7189#issue-121465814 with info from the other PR and better changelog entries.

michaelDCurran commented:
When this was merged to next, it failed a test.
```
Traceback (most recent call last):
  File "C:\Users\mick\programming\git\nvda\tests\unit\test_braille.py", line 32, in setUp
    braille.handler.handleGainFocus(self.obj)
  File "C:\Users\mick\programming\git\nvda\source\braille.py", line 1746, in handleGainFocus
    self._doNewObject(itertools.chain(getFocusContextRegions(obj, oldFocusRegions=self.mainBuffer.regions), getFocusRegi
ons(obj)))
  File "C:\Users\mick\programming\git\nvda\source\braille.py", line 1751, in _doNewObject
    for region in regions:
  File "C:\Users\mick\programming\git\nvda\source\braille.py", line 1449, in getFocusContextRegions
    region.update()
  File "C:\Users\mick\programming\git\nvda\source\braille.py", line 600, in update
    super(NVDAObjectRegion, self).update()
  File "C:\Users\mick\programming\git\nvda\source\braille.py", line 382, in update
    mode |= louis.pass1Only
AttributeError: 'module' object has no attribute 'pass1Only' 
```

leonardder commented:
hmm, it looks like the merge went wrong in that case. Liblouis pass1only has been removed as part of #7839.

michaelDCurran commented:
Right. But they are both now in Next.

I'm assuming that #7189 is making reference to passOneOnly, but  the 
Liblouis PR has removed it.

The easiest thing to do is to probably change #7189 to not reference 
passOneOnly. It does mean though that  we aare assuming that the 
libLouis pr will definitely make it to master before #7189, which I 
think is fine.

michaelDCurran commented:
Sorry, when I tested locally I had not updated my liblouis submodule it seems.  The error in the build was 
simply a missing translator comment. I shall fix myself.

michaelDCurran committed:
Add translator comment

dkager commented:
Just to confirm: this PR does not depend on any liblouis features.

michaelDCurran commented:
Yep, my mistake. Sorry about that. However, there was a missing 
translation comment, but I added it.
